typescript/sdk/src/token/CosmosNativeWarpRouteReader.ts

```diff
@@ -0,0 +1,189 @@
+import {
+  HyperlaneModuleClient,
+  SigningHyperlaneModuleClient,
+} from '@hyperlane-xyz/cosmos-sdk';
+import { warpTypes } from '@hyperlane-xyz/cosmos-types';
+import { Address, assert, rootLogger } from '@hyperlane-xyz/utils';
+
+import { CosmosNativeHookReader } from '../hook/CosmosNativeHookReader.js';
+import { CosmosNativeIsmReader } from '../ism/CosmosNativeIsmReader.js';
+import { ChainMetadataManager } from '../metadata/ChainMetadataManager.js';
+import {
+  DestinationGas,
+  MailboxClientConfig,
+  RemoteRouters,
+  RemoteRoutersSchema,
+} from '../router/types.js';
+import { ChainNameOrId, DeployedOwnableConfig } from '../types.js';
+
+import { TokenType } from './config.js';
+import { DerivedTokenRouterConfig, HypTokenConfig } from './types.js';
+
+export class CosmosNativeWarpRouteReader {
+  protected readonly logger = rootLogger.child({
+    module: 'CosmosNativeWarpRouteReader',
+  });
+  hookReader: CosmosNativeHookReader;
+  ismReader: CosmosNativeIsmReader;
+
+  constructor(
+    protected readonly metadataManager: ChainMetadataManager,
+    protected readonly chain: ChainNameOrId,
+    protected readonly cosmosProviderOrSigner:
+      | SigningHyperlaneModuleClient
+      | HyperlaneModuleClient,
+  ) {
+    this.hookReader = new CosmosNativeHookReader(
+      metadataManager,
+      cosmosProviderOrSigner,
+    );
+    this.ismReader = new CosmosNativeIsmReader(
+      metadataManager,
+      cosmosProviderOrSigner,
+    );
+  }
+
+  /**
+   * Derives the configuration for a Hyperlane ERC20 router contract at the given address.
+   *
+   * @param warpRouteAddress - The address of the Hyperlane ERC20 router contract.
+   * @returns The configuration for the Hyperlane ERC20 router.
+   *
+   */
+  async deriveWarpRouteConfig(
+    warpRouteAddress: Address,
+  ): Promise<DerivedTokenRouterConfig> {
+    // Derive the config type
+    const type = await this.deriveTokenType(warpRouteAddress);
+    const baseMetadata = await this.fetchMailboxClientConfig(warpRouteAddress);
+    const tokenConfig = await this.fetchTokenConfig(type, warpRouteAddress);
+    const remoteRouters = await this.fetchRemoteRouters(warpRouteAddress);
+    const proxyAdmin = await this.fetchProxyAdminConfig(warpRouteAddress);
+    const destinationGas = await this.fetchDestinationGas(warpRouteAddress);
```

Comment:
Would it be possible to `Promise.all` or there would be concerns with rpc rate limiting?

Reply:
this could lead to issues with ratelimiting. Most rpc endpoints really don't like parallel requests